Error linking with libquadmath

I kept getting linking errors for a while, and I just noticed that it's also causing issues building brew bottles on ubuntu:
```
/home/linuxbrew/.linuxbrew/opt/binutils/bin/ld: Macaulay2/d/CMakeFiles/M2-interpreter.dir/boostmath-tmp.cc.o: undefined reference to symbol 'ldexpq@@QUADMATH_1.0'
/home/linuxbrew/.linuxbrew/opt/binutils/bin/ld: /home/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current/libquadmath.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
On my machine linking with `quadmath` resolves this, but apparently this fails on MacOS. @d-torrance any ideas what might be behind this?

--- a/M2/Macaulay2/d/CMakeLists.txt
+++ b/M2/Macaulay2/d/CMakeLists.txt
@@ -147,6 +147,11 @@ add_library(M2-interpreter OBJECT ${CLIST} ${CXXLIST} ${TAGS}
 target_link_libraries(M2-interpreter PRIVATE M2-supervisor Boost::regex TBB::tbb
   $<$<BOOL:${WITH_FFI}>:FFI::ffi>)
 
+# TODO: c.f. https://github.com/Macaulay2/M2/issues/2682
+if(NOT APPLE)
+  target_link_libraries(M2-interpreter PRIVATE quadmath)
+endif()
+
 target_include_directories(M2-interpreter
   PUBLIC
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>

--- a/M2/Macaulay2/e/Eschreyer.cpp
+++ b/M2/Macaulay2/e/Eschreyer.cpp
@@ -314,13 +314,13 @@ void GBKernelComputation::wipe_unneeded_terms(gbvector *&f)
   // Remove every term of f (except the lead term)
   // which is NOT divisible by an element of mi.
   int *exp = newarray_atomic(int, GR->n_vars());
-  int nterms = 1;
-  int nsaved = 0;
+  // int nterms = 1;
+  // int nsaved = 0;
   gbvector *g = f;
   while (g->next != 0)
     {
       // First check to see if the term g->next is in the monideal
-      nterms++;
+      // nterms++;
       Bag *b;
       GR->gbvector_get_lead_exponents(F, g->next, exp);
       if (mi[g->next->comp - 1]->search_expvector(exp, b))
@@ -331,7 +331,7 @@ void GBKernelComputation::wipe_unneeded_terms(gbvector *&f)
       else
         {
           // Want to dump this term
-          nsaved++;
+          // nsaved++;
           gbvector *tmp = g->next;
           g->next = tmp->next;
           tmp->next = 0;

--- a/M2/Macaulay2/e/NCAlgebras/FreeAlgebra.cpp
+++ b/M2/Macaulay2/e/NCAlgebras/FreeAlgebra.cpp
@@ -555,7 +555,7 @@ void FreeAlgebra::mult_by_term_right(Poly& result,
                                      const ring_elem c,
                                      const Word& w) const
 {
-  IntVector tmp;
+  gc_vector<int> tmp;
   monoid().monomInsertFromWord(tmp,w);
   Monom tmpMonom(tmp.data());
   mult_by_term_right(result,f,c,tmpMonom);
@@ -586,7 +586,7 @@ void FreeAlgebra::mult_by_term_left(Poly& result,
                                     const ring_elem c,
                                     const Word& w) const
 {
-  IntVector tmp;
+  gc_vector<int> tmp;
   monoid().monomInsertFromWord(tmp,w);
   Monom tmpMonom(tmp.data());
   mult_by_term_left(result,f,c,tmpMonom);
@@ -633,7 +633,7 @@ void FreeAlgebra::mult_by_term_left_and_right(Poly& result,
                                               const Word& leftW,
                                               const Word& rightW) const
 {
-  IntVector leftTmp, rightTmp;
+  gc_vector<int> leftTmp, rightTmp;
   monoid().monomInsertFromWord(leftTmp,leftW);
   monoid().monomInsertFromWord(rightTmp,rightW);
   Monom leftTmpMonom(leftTmp.data());
@@ -648,7 +648,7 @@ void FreeAlgebra::mult_by_term_left_and_right(Poly& result,
                                               const Word& leftW,
                                               const Word& rightW) const
 {
-  IntVector leftTmp, rightTmp;
+  gc_vector<int> leftTmp, rightTmp;
   monoid().monomInsertFromWord(leftTmp,leftW);
   monoid().monomInsertFromWord(rightTmp,rightW);
   Monom leftTmpMonom(leftTmp.data());

--- a/M2/Macaulay2/e/NCAlgebras/FreeMonoid.hpp
+++ b/M2/Macaulay2/e/NCAlgebras/FreeMonoid.hpp
@@ -1,8 +1,8 @@
 #ifndef _free_monoid_hpp_
 #define _free_monoid_hpp_
 
-#include "Polynomial.hpp"      // for Monom, IntVector
-#include "newdelete.hpp"       // for VECTOR, our_new_delete
+#include "Polynomial.hpp"      // for Monom
+#include "newdelete.hpp"       // for our_new_delete
 #include "polyring.hpp"        // for PolynomialRing
 #include "style.hpp"           // for GT
 #include "NCAlgebras/Word.hpp" // for Word
@@ -56,7 +56,7 @@ private:
   const std::vector<int> mHeftVector;    // length is size of degree vector
   std::vector<int> mHeftDegrees;   // length numVars().  Should be const (after construction)
   const int mNumWeights;
-  VECTOR(const int*) mDegreeOfVar;
+  gc_vector<const int*> mDegreeOfVar;
      // length numVars(), each is a pointer to an allocated degree vector.
      // Should be const (after construction)
 public:
@@ -79,7 +79,7 @@ public:
   const Monoid& degreeMonoid() const { return * mDegreeRing->getMonoid(); }
   
   // Monomial operations
-  using MonomialInserter = IntVector;
+  using MonomialInserter = gc_vector<int>;
 
   void one(MonomialInserter& m) const;
 

--- a/M2/Macaulay2/e/Polynomial.hpp
+++ b/M2/Macaulay2/e/Polynomial.hpp
@@ -1,7 +1,7 @@
 #ifndef _polynomial_hpp_
 #define _polynomial_hpp_
 
-#include "newdelete.hpp"  // for VECTOR, our_new_delete
+#include "newdelete.hpp"  // for our_new_delete
 #include "ringelem.hpp"   // for ring_elem
 #include "style.hpp"      // for GT, LT, EQ
 
@@ -10,9 +10,6 @@
 #include <iostream>       // for ostream
 #include <iterator>       // for forward_iterator_tag
 #include <utility>        // for pair, make_pair
-
-using IntVector = VECTOR(int);
-// using IntVector = std::vector<int>;
 
 struct Monom
 // Format for monomials:
@@ -158,8 +155,8 @@ class Polynomial : public our_new_delete
   
   typedef typename CoefficientRingType::ElementType ElementType;
 public:  
-  typedef typename VECTOR(ElementType) coeffVector;
-  using monomVector = IntVector; // TODO: remove monomVector?
+  using coeffVector = gc_vector<ElementType>;
+  using monomVector = gc_vector<int>; // TODO: remove monomVector?
 
   typedef typename coeffVector::iterator coeffIterator;
   typedef monomVector::iterator monomIterator;
@@ -264,8 +261,8 @@ struct CoefficientRingType
 };
 
 using Poly = Polynomial<CoefficientRingType>;
-using PolyList = VECTOR(Poly*);
-using ConstPolyList = VECTOR(const Poly*);
+using PolyList = gc_vector<Poly*>;
+using ConstPolyList = gc_vector<const Poly*>;
 
 #endif
 

--- a/M2/Macaulay2/e/newdelete.hpp
+++ b/M2/Macaulay2/e/newdelete.hpp
@@ -29,7 +29,10 @@
   It is a (probably hard to detect) error if exactly (1) or (2a) holds.
   In that case, redesign your class!
  */
-#define VECTOR(T) std::vector<T, gc_allocator<T>>
+template <typename T>
+using gc_vector = typename std::vector<T, gc_allocator<T>>;
+// TODO: eventually replace all uses of VECTOR(T) with gc_vector<T>
+#define VECTOR(T) gc_vector<T>
 
 // these replace all uses of the construction "new T[n]" (unless constructors
 // have to be run!):

--- a/M2/Macaulay2/e/newdelete.hpp
+++ b/M2/Macaulay2/e/newdelete.hpp
@@ -1,18 +1,19 @@
 #ifndef NEWDELETE_H
 #define NEWDELETE_H 1
 
-#define GC_REDIRECT_TO_LOCAL
-// get declarations of outofmem and getmem
-#include "../d/M2mem.h"
-#include "../d/debug.h"
+#define GC_REDIRECT_TO_LOCAL // enable thread-local allocation
 
-#include <M2/gc-include.h>
+#include "M2mem.h"  // for freemem, getmem, outofmem2
+#include "debug.h"  // for TRAPCHK, TRAPCHK_SIZE
+
+#include "M2/gc-include.h"  // for GC_base, NULL, size_t, GC_REGISTER_FINALI...
+// IWYU pragma: begin_exports
 #include <gc/gc_allocator.h>
+// IWYU pragma: end_exports
 #include <vector>
 
 #ifdef MEMDEBUG
 #include <memdebug.h>
-#include <M2mem.h>
 #endif
 
 /**

--- a/M2/Macaulay2/e/unit-tests/M2mem-replacement.c
+++ b/M2/Macaulay2/e/unit-tests/M2mem-replacement.c
@@ -34,9 +34,8 @@ void outofmem(void) {
 
 void outofmem2(size_t new) {
      const char *msg = "\n\n *** out of memory trying to allocate %ld bytes, exiting ***\n";
-     const int N = sizeof(msg) + 100;
-     static char buf[N];
-     snprintf(buf,N,msg,(long)new);
+     static char buf[sizeof(msg) + 100];
+     sprintf(buf,msg,(long)new);
      int r = write(STDERR,buf,strlen(buf));
      if (r == ERROR) exit(1);
      exit(1);

